### PR TITLE
Refactor memory wrappers

### LINKED ISSUE #875: CuArrays don't seem to display correctly in VS code
**Describe the bug**

Seems like CuArrays don't display correctly in the VS code REPL on windows, although the values seem to be ok. Don't have a linux install to test on. Also seems to work fine when running julia in windows terminal.

**To reproduce**

The Minimal Working Example (MWE) for this bug:

```julia
julia> xx = randn(3)
3-element Vector{Float64}:
 -0.2691051505932
 -0.38601362073983303
 -0.16102191893287385

julia> cu(xx)
3-element CuArray{Float32, 1}:
 0.0
 0.0
 0.0

julia> cu(xx) |> Array
3-element Vector{Float32}:
 -0.26910514
 -0.38601363
 -0.16102192

julia> cu(xx) .+ 10 |> Array
3-element Vector{Float32}:
 9.730895
 9.613986
 9.838978

# These seem to work, VS code uses some special MIME or something?
julia> display(cu(xx))
3-element CuArray{Float32, 1}:
 -0.26910514
 -0.38601363
 -0.16102192

julia> show(cu(xx))
Float32[-0.26910514, -0.38601363, -0.16102192]

# This is a bit puzzling...
julia> @show cu(xx)
cu(xx) = Float32[-0.26910514, -0.38601363, -0.16102192]
3-element CuArray{Float32, 1}:
 -0.26910514
 -0.38601363
 -0.16102192

# These are not 100% reproducible. Threading issue?
julia> cu(xx) |> identity
3-element CuArray{Float32, 1}:
 -0.26910514
 -0.38601363
 -0.16102192

julia> cu(xx) .+ 10
3-element CuArray{Float32, 1}:
 -0.26910514
 -0.38601363
 -0.16102192
```

<details><summary>Manifest.toml</summary>
<p>

```
# This file is machine-generated - editing it directly is not advised

[[AbstractFFTs]]
deps = ["LinearAlgebra"]
git-tree-sha1 = "485ee0867925449198280d4af84bdb46a2a404d0"
uuid = "621f4979-c628-5d54-868e-fcf4e3e8185c"
version = "1.0.1"

[[Adapt]]
deps = ["LinearAlgebra"]
git-tree-sha1 = "f1b523983a58802c4695851926203b36e28f09db"
uuid = "79e6a3ab-5dfb-504d-930d-738a2a938a0e"
version = "3.3.0"

[[ArgTools]]
uuid = "0dad84c5-d112-42e6-8d28-ef12dabb789f"

[[Artifacts]]
uuid = "56f22d72-fd6d-98f1-02f0-08ddc0907c33"

[[BFloat16s]]
deps = ["LinearAlgebra", "Test"]
git-tree-sha1 = "4af69e205efc343068dc8722b8dfec1ade89254a"
uuid = "ab4f0b2a-ad5b-11e8-123f-65d77653426b"
version = "0.1.0"

[[Base64]]
uuid = "2a0f44e3-6c83-55bd-87e4-b1978d98bd5f"

[[CEnum]]
git-tree-sha1 = "215a9aa4a1f23fbd05b92769fdd62559488d70e9"
uuid = "fa961155-64e5-5f13-b03f-caf6b980ea82"
version = "0.4.1"

[[CUDA]]
deps = ["AbstractFFTs", "Adapt", "BFloat16s", "CEnum", "CompilerSupportLibraries_jll", "DataStructures", "ExprTools", "GPUArrays", "GPUCompiler", "LLVM", "LazyArtifacts", "Libdl", "LinearAlgebra", "Logging", "MacroTools", "Memoize", "Printf", "Random", "RandomNumbers", "Reexport", "Requires", "SparseArrays", "SpecialFunctions", "Statistics", "TimerOutputs"]
git-tree-sha1 = "d4fa6486e94c4087f1d081d7be2d501a170bd51d"
uuid = "052768ef-5323-5732-b1bb-66c8b64840ba"
version = "3.1.0"

[[ChainRulesCore]]
deps = ["Compat", "LinearAlgebra", "SparseArrays"]
git-tree-sha1 = "a66109c73612c63b10923ac446fddb0f0d21a593"
uuid = "d360d2e6-b24c-11e9-a2a3-2a2ae2dbcce4"
version = "0.9.40"

[[Compat]]
deps = ["Base64", "Dates", "DelimitedFiles", "Distributed", "InteractiveUtils", "LibGit2", "Libdl", "LinearAlgebra", "Markdown", "Mmap", "Pkg", "Printf", "REPL", "Random", "SHA", "Serialization", "SharedArrays", "Sockets", "SparseArrays", "Statistics", "Test", "UUIDs", "Unicode"]
git-tree-sha1 = "ac4132ad78082518ec2037ae5770b6e796f7f956"
uuid = "34da2185-b29b-5c13-b0c7-acf172513d20"
version = "3.27.0"

[[CompilerSupportLibraries_jll]]
deps = ["Artifacts", "Libdl"]
uuid = "e66e0078-7015-5450-92f7-15fbd957f2ae"

[[DataStructures]]
deps = ["Compat", "InteractiveUtils", "OrderedCollections"]
git-tree-sha1 = "4437b64df1e0adccc3e5d1adbc3ac741095e4677"
uuid = "864edb3b-99cc-5e75-8d2d-829cb0a9cfe8"
version = "0.18.9"

[[Dates]]
deps = ["Printf"]
uuid = "ade2ca70-3891-5945-98fb-dc099432e06a"

[[DelimitedFiles]]
deps = ["Mmap"]
uuid = "8bb1440f-4735-579b-a4ab-409b98df4dab"

[[Distributed]]
deps = ["Random", "Serialization", "Sockets"]
uuid = "8ba89e20-285c-5b6f-9357-94700520ee1b"

[[Downloads]]
deps = ["ArgTools", "LibCURL", "NetworkOptions"]
uuid = "f43a241f-c20a-4ad4-852c-f6b1247861c6"

[[ExprTools]]
git-tree-sha1 = "10407a39b87f29d47ebaca8edbc75d7c302ff93e"
uuid = "e2ba6199-217a-4e67-a87a-7c52f15ade04"
version = "0.1.3"

[[GPUArrays]]
deps = ["AbstractFFTs", "Adapt", "LinearAlgebra", "Printf", "Random", "Serialization"]
git-tree-sha1 = "9c95b2fd5c16bc7f97371e9f92f0fef77e0f5957"
uuid = "0c68f7d7-f131-5f86-a1c3-88cf8149b2d7"
version = "6.2.2"

[[GPUCompiler]]
deps = ["DataStructures", "ExprTools", "InteractiveUtils", "LLVM", "Libdl", "Logging", "Scratch", "Serialization", "TimerOutputs", "UUIDs"]
git-tree-sha1 = "6eadd2321dc3ac0fc9d530ab01c2caa7fe5d74c6"
uuid = "61eb1bfa-7361-4325-ad38-22787b887f55"
version = "0.11.4"

[[InteractiveUtils]]
deps = ["Markdown"]
uuid = "b77e0a4c-d291-57a0-90e8-8db25a27a240"

[[JLLWrappers]]
deps = ["Preferences"]
git-tree-sha1 = "642a199af8b68253517b80bd3bfd17eb4e84df6e"
uuid = "692b3bcd-3c85-4b1f-b108-f13ce0eb3210"
version = "1.3.0"

[[LLVM]]
deps = ["CEnum", "Libdl", "Printf", "Unicode"]
git-tree-sha1 = "b616937c31337576360cb9fb872ec7633af7b194"
uuid = "929cbde3-209d-540e-8aea-75f648917ca0"
version = "3.6.0"

[[LazyArtifacts]]
deps = ["Artifacts", "Pkg"]
uuid = "4af54fe1-eca0-43a8-85a7-787d91b784e3"

[[LibCURL]]
deps = ["LibCURL_jll", "MozillaCACerts_jll"]
uuid = "b27032c2-a3e7-50c8-80cd-2d36dbcbfd21"

[[LibCURL_jll]]
deps = ["Artifacts", "LibSSH2_jll", "Libdl", "MbedTLS_jll", "Zlib_jll", "nghttp2_jll"]
uuid = "deac9b47-8bc7-5906-a0fe-35ac56dc84c0"

[[LibGit2]]
deps = ["Base64", "NetworkOptions", "Printf", "SHA"]
uuid = "76f85450-5226-5b5a-8eaa-529ad045b433"

[[LibSSH2_jll]]
deps = ["Artifacts", "Libdl", "MbedTLS_jll"]
uuid = "29816b5a-b9ab-546f-933c-edad1886dfa8"

[[Libdl]]
uuid = "8f399da3-3557-5675-b5ff-fb832c97cbdb"

[[LinearAlgebra]]
deps = ["Libdl"]
uuid = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"

[[Logging]]
uuid = "56ddb016-857b-54e1-b83d-db4d58db5568"

[[MacroTools]]
deps = ["Markdown", "Random"]
git-tree-sha1 = "6a8a2a625ab0dea913aba95c11370589e0239ff0"
uuid = "1914dd2f-81c6-5fcd-8719-6d5c9610ff09"
version = "0.5.6"

[[Markdown]]
deps = ["Base64"]
uuid = "d6f4376e-aef5-505a-96c1-9c027394607a"

[[MbedTLS_jll]]
deps = ["Artifacts", "Libdl"]
uuid = "c8ffd9c3-330d-5841-b78e-0817d7145fa1"

[[Memoize]]
deps = ["MacroTools"]
git-tree-sha1 = "2b1dfcba103de714d31c033b5dacc2e4a12c7caa"
uuid = "c03570c3-d221-55d1-a50c-7939bbd78826"
version = "0.4.4"

[[Mmap]]
uuid = "a63ad114-7e13-5084-954f-fe012c677804"

[[MozillaCACerts_jll]]
uuid = "14a3606d-f60d-562e-9121-12d972cd8159"

[[NetworkOptions]]
uuid = "ca575930-c2e3-43a9-ace4-1e988b2c1908"

[[OpenSpecFun_jll]]
deps = ["Artifacts", "CompilerSupportLibraries_jll", "JLLWrappers", "Libdl", "Pkg"]
git-tree-sha1 = "b9b8b8ed236998f91143938a760c2112dceeb2b4"
uuid = "efe28fd5-8261-553b-a9e1-b2916fc3738e"
version = "0.5.4+0"

[[OrderedCollections]]
git-tree-sha1 = "4fa2ba51070ec13fcc7517db714445b4ab986bdf"
uuid = "bac558e1-5e72-5ebc-8fee-abe8a469f55d"
version = "1.4.0"

[[Pkg]]
deps = ["Artifacts", "Dates", "Downloads", "LibGit2", "Libdl", "Logging", "Markdown", "Printf", "REPL", "Random", "SHA", "Serialization", "TOML", "Tar", "UUIDs", "p7zip_jll"]
uuid = "44cfe95a-1eb2-52ea-b672-e2afdf69b78f"

[[Preferences]]
deps = ["TOML"]
git-tree-sha1 = "ea79e4c9077208cd3bc5d29631a26bc0cff78902"
uuid = "21216c6a-2e73-6563-6e65-726566657250"
version = "1.2.1"

[[Printf]]
deps = ["Unicode"]
uuid = "de0858da-6303-5e67-8744-51eddeeeb8d7"

[[REPL]]
deps = ["InteractiveUtils", "Markdown", "Sockets", "Unicode"]
uuid = "3fa0cd96-eef1-5676-8a61-b3b8758bbffb"

[[Random]]
deps = ["Serialization"]
uuid = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"

[[RandomNumbers]]
deps = ["Random", "Requires"]
git-tree-sha1 = "441e6fc35597524ada7f85e13df1f4e10137d16f"
uuid = "e6cf234a-135c-5ec9-84dd-332b85af5143"
version = "1.4.0"

[[Reexport]]
git-tree-sha1 = "57d8440b0c7d98fc4f889e478e80f268d534c9d5"
uuid = "189a3867-3050-52da-a836-e630ba90ab69"
version = "1.0.0"

[[Requires]]
deps = ["UUIDs"]
git-tree-sha1 = "4036a3bd08ac7e968e27c203d45f5fff15020621"
uuid = "ae029012-a4dd-5104-9daa-d747884805df"
version = "1.1.3"

[[SHA]]
uuid = "ea8e919c-243c-51af-8825-aaa63cd721ce"

[[Scratch]]
deps = ["Dates"]
git-tree-sha1 = "ad4b278adb62d185bbcb6864dc24959ab0627bf6"
uuid = "6c6a2e73-6563-6170-7368-637461726353"
version = "1.0.3"

[[Serialization]]
uuid = "9e88b42a-f829-5b0c-bbe9-9e923198166b"

[[SharedArrays]]
deps = ["Distributed", "Mmap", "Random", "Serialization"]
uuid = "1a1011a3-84de-559e-8e89-a11a2f7dc383"

[[Sockets]]
uuid = "6462fe0b-24de-5631-8697-dd941f90decc"

[[SparseArrays]]
deps = ["LinearAlgebra", "Random"]
uuid = "2f01184e-e22b-5df5-ae63-d93ebab69eaf"

[[SpecialFunctions]]
deps = ["ChainRulesCore", "OpenSpecFun_jll"]
git-tree-sha1 = "5919936c0e92cff40e57d0ddf0ceb667d42e5902"
uuid = "276daf66-3868-5448-9aa4-cd146d93841b"
version = "1.3.0"

[[Statistics]]
deps = ["LinearAlgebra", "SparseArrays"]
uuid = "10745b16-79ce-11e8-11f9-7d13ad32a3b2"

[[TOML]]
deps = ["Dates"]
uuid = "fa267f1f-6049-4f14-aa54-33bafae1ed76"

[[Tar]]
deps = ["ArgTools", "SHA"]
uuid = "a4e569a6-e804-4fa4-b0f3-eef7a1d5b13e"

[[Test]]
deps = ["InteractiveUtils", "Logging", "Random", "Serialization"]
uuid = "8dfed614-e22c-5e08-85e1-65c5234f0b40"

[[TimerOutputs]]
deps = ["Printf"]
git-tree-sha1 = "32cdbe6cd2d214c25a0b88f985c9e0092877c236"
uuid = "a759f4b9-e2f1-59dc-863e-4aeb61b1ea8f"
version = "0.5.8"

[[UUIDs]]
deps = ["Random", "SHA"]
uuid = "cf7118a7-6976-5b1a-9a39-7adc72f591a4"

[[Unicode]]
uuid = "4ec0a83e-493e-50e2-b9ac-8f72acf5a8f5"

[[Zlib_jll]]
deps = ["Libdl"]
uuid = "83775a58-1f1d-513f-b197-d71354ab007a"

[[nghttp2_jll]]
deps = ["Artifacts", "Libdl"]
uuid = "8e850ede-7688-5339-a07c-302acd2aaf8d"

[[p7zip_jll]]
deps = ["Artifacts", "Libdl"]
uuid = "3f19e933-33d8-53b3-aaab-bd5110c3b7a0"

```

</p>
</details>

**Expected behavior**

Values should be represented correctly.

**Version info**

Details on Julia:

```
# please post the output of:
julia> versioninfo()
Julia Version 1.6.1
Commit 6aaedecc44 (2021-04-23 05:59 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i7-5820K CPU @ 3.30GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, haswell)
Environment:
  JULIA_DEPOT_PATH = E:/Programs/julia/.julia
  JULIA_EDITOR = code
  JULIA_NUM_THREADS = 12
```

Details on CUDA:

```
# please post the output of:
julia> CUDA.version()
v"11.3.0"
```

**Additional context**

### LINKED ISSUE #1566: Docs: add example on task-based parallelism with explicit synchronization
The current example calls `Array` in the `@async` block, which is implicitly synchronizing. The docs should probably explain how that works, and that you want to use an explicit `synchronize()` if you want to keep the GPU memory.

### LINKED ISSUE #2216: LazyInitialized is not GC-safe
Yet another issue caused by finalizers not running on their own thread:

```
Stacktrace:
  [1] try_yieldto(undo::typeof(Base.ensure_rescheduled))
    @ Base ./task.jl:921
  [2] wait()
    @ Base ./task.jl:995
  [3] yield()
    @ Base ./task.jl:875
  [4] initialize!(x::CUDA.APIUtils.LazyInitialized{Vector{Vector{Union{Nothing, Bool}}}, CUDA.var"#1003#1004"}, constructor::CUDA.var"#1005#1006")
    @ CUDA.APIUtils ~/Julia/pkg/CUDA/lib/utils/threading.jl:57
  [5] get!
    @ CUDA.APIUtils ~/Julia/pkg/CUDA/lib/utils/threading.jl:28 [inlined]
  [6] macro expansion
    @ CUDA ~/Julia/pkg/CUDA/lib/utils/memoization.jl:69 [inlined]
  [7] stream_ordered(dev::CuDevice)
    @ CUDA ~/Julia/pkg/CUDA/src/pool.jl:163
  [8] #_free#1023
    @ ~/Julia/pkg/CUDA/src/pool.jl:510 [inlined]
  [9] _free
    @ ~/Julia/pkg/CUDA/src/pool.jl:503 [inlined]
 [10] macro expansion
    @ ~/Julia/pkg/CUDA/src/pool.jl:488 [inlined]
 [11] macro expansion
    @ ./timing.jl:395 [inlined]
 [12] #free#1022
    @ ~/Julia/pkg/CUDA/src/pool.jl:487 [inlined]
 [13] free
    @ ~/Julia/pkg/CUDA/src/pool.jl:476 [inlined]
 [14] (::CUDA.var"#1029#1030"{CUDA.Mem.DeviceBuffer, Bool})()
    @ CUDA ~/Julia/pkg/CUDA/src/array.jl:101
 [15] #context!#912
    @ ~/Julia/pkg/CUDA/lib/cudadrv/state.jl:170 [inlined]
 [16] context!(ctx::CuContext)
    @ CUDA ~/Julia/pkg/CUDA/lib/cudadrv/state.jl:165 [inlined]
 [17] _free_buffer(buf::CUDA.Mem.DeviceBuffer, early::Bool)
    @ CUDA ~/Julia/pkg/CUDA/src/array.jl:89
 [18] release(rc::GPUArrays.RefCounted{CUDA.Mem.DeviceBuffer}, args::Bool)
    @ GPUArrays ~/.julia/packages/GPUArrays/EoKy0/src/host/abstractarray.jl:42
 [19] unsafe_free!
    @ GPUArrays ~/.julia/packages/GPUArrays/EoKy0/src/host/abstractarray.jl:90 [inlined]
 [20] unsafe_finalize!(xs::CuArray{Float16, 2, CUDA.Mem.DeviceBuffer})
    @ CUDA ~/Julia/pkg/CUDA/src/array.jl:113
```

### LINKED ISSUE #2236: Track CuArray stream usage
We currently require that tasks where CuArrays are written to, synchronize the GPU before sending that data off to other tasks. This is fragile, difficult to explain, and a source of common errors. We should consider tracking which tasks or streams a CuArray is used in, in order to automatically synchronize on use.

### LINKED ISSUE #2323: Improve cross-device usage
I discovered this very surprising behavior where performing operations between arrays on a device produces an array on a different device and then I also get an error:
```julia
julia> using CUDA

julia> CUDA.device!(0)
CuDevice(0): NVIDIA TITAN RTX

julia> x = CUDA.ones(2)
2-element CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}:
 1.0
 1.0

julia> CUDA.device!(1)
CuDevice(1): NVIDIA TITAN RTX

julia> x
2-element CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}:
 1.0
 1.0

julia> y = x - x;

julia> CUDA.device(x)
CuDevice(0): NVIDIA TITAN RTX

julia> CUDA.device(y)
CuDevice(1): NVIDIA TITAN RTX

julia> y
2-element CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}:
Error showing value of type CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}:
ERROR: CUDA error: an illegal memory access was encountered (code 700, ERROR_ILLEGAL_ADDRESS)
Stacktrace:
  [1] throw_api_error(res::CUDA.cudaError_enum)
    @ CUDA ~/.julia/packages/CUDA/fGE8R/lib/cudadrv/libcuda.jl:30
  [2] isdone
    @ ~/.julia/packages/CUDA/fGE8R/lib/cudadrv/stream.jl:111 [inlined]
  [3] spinning_synchronization(f::typeof(CUDA.isdone), obj::CuStream)
    @ CUDA ~/.julia/packages/CUDA/fGE8R/lib/cudadrv/synchronization.jl:79
  [4] synchronize(stream::CuStream; blocking::Bool, spin::Bool)
    @ CUDA ~/.julia/packages/CUDA/fGE8R/lib/cudadrv/synchronization.jl:196
  [5] synchronize (repeats 2 times)
    @ ~/.julia/packages/CUDA/fGE8R/lib/cudadrv/synchronization.jl:194 [inlined]
  [6] (::CUDA.var"#1098#1099"{Float32, Vector{Float32}, Int64, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, Int64, Int64})()
    @ CUDA ~/.julia/packages/CUDA/fGE8R/src/array.jl:606
  [7] #context!#954
    @ ~/.julia/packages/CUDA/fGE8R/lib/cudadrv/state.jl:170 [inlined]
  [8] context!
    @ ~/.julia/packages/CUDA/fGE8R/lib/cudadrv/state.jl:165 [inlined]
  [9] unsafe_copyto!(dest::Vector{Float32}, doffs::Int64, src::CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, soffs::Int64, n::Int64)
    @ CUDA ~/.julia/packages/CUDA/fGE8R/src/array.jl:602
 [10] copyto!
    @ ~/.julia/packages/CUDA/fGE8R/src/array.jl:555 [inlined]
 [11] copyto!
    @ ~/.julia/packages/CUDA/fGE8R/src/array.jl:559 [inlined]
 [12] copyto_axcheck!
    @ ./abstractarray.jl:1177 [inlined]
 [13] Vector{Float32}(x::CuArray{Float32, 1, CUDA.Mem.DeviceBuffer})
    @ Base ./array.jl:673
 [14] Array
    @ ./boot.jl:500 [inlined]
 [15] convert
    @ ./array.jl:665 [inlined]
 [16] adapt_storage
    @ ~/.julia/packages/GPUArrays/OKkAu/src/host/abstractarray.jl:115 [inlined]
 [17] adapt_structure
    @ ~/.julia/packages/Adapt/7T9au/src/Adapt.jl:57 [inlined]
 [18] adapt
    @ ~/.julia/packages/Adapt/7T9au/src/Adapt.jl:40 [inlined]
 [19] print_array
    @ ~/.julia/packages/GPUArrays/OKkAu/src/host/abstractarray.jl:118 [inlined]
 [20] show(io::IOContext{Base.TTY}, ::MIME{Symbol("text/plain")}, X::CuArray{Float32, 1, CUDA.Mem.DeviceBuffer})
    @ Base ./arrayshow.jl:399
``` 

```julia
julia> CUDA.versioninfo()
CUDA runtime 12.4, artifact installation
CUDA driver 12.0
NVIDIA driver 525.89.2

CUDA libraries: 
- CUBLAS: 12.4.5
- CURAND: 10.3.5
- CUFFT: 11.2.1
- CUSOLVER: 11.6.1
- CUSPARSE: 12.3.1
- CUPTI: 22.0.0
- NVML: 12.0.0+525.89.2

Julia packages: 
- CUDA: 5.3.0
- CUDA_Driver_jll: 0.8.1+0
- CUDA_Runtime_jll: 0.12.1+0

Toolchain:
- Julia: 1.10.2
- LLVM: 15.0.7

3 devices:
  0: NVIDIA TITAN RTX (sm_75, 13.265 GiB / 24.000 GiB available)
  1: NVIDIA TITAN RTX (sm_75, 23.447 GiB / 24.000 GiB available)
  2: NVIDIA TITAN RTX (sm_75, 23.445 GiB / 24.000 GiB available)
```

### PR DESCRIPTION
**Pull request by maleadt**
Implements https://github.com/JuliaGPU/CUDA.jl/issues/2236: Each piece of memory now keeps track on which stream is working on it, and whether there's outstanding (non-synchronized) changes within that stream. That makes it possible to accurately and automatically synchronize when we derive a pointer from a CuArray (which is the one and only place where we indicate that we will start using the memory).

I did quite a bit of refactoring as part of this PR, most notably removing the `Mem` submodule. I added plenty deprecations though, so it should be nonbreaking (although I expect issues and hotfixes when we tag this).

---

Closes https://github.com/JuliaGPU/CUDA.jl/issues/1566 by making synchronization unnecessary, i.e., the following now "just works":

```julia
a = CUDA.zeros(4096, 4096)
b = CUDA.zeros(4096, 4096)
t = @async begin
    # PREVIOUSLY WRONG: using `a` while it may not be ready
    for x in 1:100
        mul!(b, a, a)
    end
    a .= 1
end
wait(t)
# PREVIOUSLY WRONG: using `a` while it may not be ready
display(Array(a)[end])
```

@vchuravy: much easier than the Task changes we were brainstorming about :-)

---

Fixes https://github.com/JuliaGPU/CUDA.jl/issues/2323:

```julia
julia> using CUDA

julia> CUDA.device!(0)
CuDevice(0): Tesla V100-PCIE-16GB

julia> x = CUDA.ones(2)

2-element CuArray{Float32, 1, CUDA.DeviceMemory}:
 1.0
 1.0

julia> CUDA.device!(1)
CuDevice(1): Tesla V100S-PCIE-32GB

julia> x
2-element CuArray{Float32, 1, CUDA.DeviceMemory}:
 1.0
 1.0

julia> y = x - x;

julia> CUDA.device(x)
CuDevice(0): Tesla V100-PCIE-16GB

julia> CUDA.device(y)
CuDevice(1): Tesla V100S-PCIE-32GB

julia> y
2-element CuArray{Float32, 1, CUDA.DeviceMemory}:
 0.0
 0.0
```

---

Hopefully fixes https://github.com/JuliaGPU/CUDA.jl/issues/875

### CODE REVIEW

File: lib/cudadrv/stream.jl [hunk 1]
```diff
@@ -11,6 +11,8 @@ Create a CUDA stream.
 """
 mutable struct CuStream
     handle::CUstream
+    valid::Bool
```
Comment (vchuravy) [inline lib/cudadrv/stream.jl]: Should this be `@atomic`? And the handle `const`

Comment (maleadt) [reply lib/cudadrv/stream.jl]: The validity is only set from a finalizer, does that warrant making the field atomic?

Comment (vchuravy) [reply lib/cudadrv/stream.jl]: Technically possible to have more than one finalizer attached to an object and eventually possible that they might race each other, but currently no.

Comment (maleadt) [reply lib/cudadrv/stream.jl]: But still a good suggestion, I'll add it.

**Pull request merged**